tidy3d/plugins/design/method.py

```diff
@@ -175,28 +110,466 @@ class MethodGrid(MethodIndependent):
+                console.log(
+                    f"Generation {ga_instance.generations_completed} Best Fitness: {best_fitness:.3f}"
+                )
+
+        # Determine initial array
+        num_genes = len(parameters)
+
+        # define the optimizer
+        ga_instance = pygad.GA(
+            num_generations=self.n_generations,
+            num_parents_mating=self.n_parents_mating,
+            fitness_func=fitness_function,
+            parent_selection_type=self.parent_selection_type,
+            mutation_type=self.mutation_type,
+            mutation_probability=self.mutation_prob,
+            crossover_type=self.crossover_type,
+            crossover_probability=self.crossover_prob,
+            sol_per_pop=self.solutions_per_pop,
+            num_genes=num_genes,
+            fitness_batch_size=self.solutions_per_pop,
+            on_generation=on_generation,
+            random_seed=self.rng_seed,
+            gene_space=gene_spaces,
+            gene_type=gene_types,
+            stop_criteria=self.stop_criteria,
+            save_solutions=True,
+        )
+
+        ga_instance.run()
+
+        if console is not None:
+            solution, solution_fitness, _ = ga_instance.best_solution()
+            console.log(
+                f"Best Result: {solution_fitness}\n"
+                f"Best Parameters: {' '.join([f'{param.name}: {value}' for param, value in zip(parameters, solution)])}\n"
+            )
+
+        # Format output
+        fn_args = [dict(zip(param_keys, val)) for arr in store_parameters for val in arr]
+        results = [val for arr in store_fitness for val in arr]
+
+        return fn_args, results, store_aux, ga_instance
+
+
+class MethodParticleSwarm(MethodOptimize, ABC):
+    """A standard method for performing particle swarm search"""
+
+    n_particles: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of particles in the swarm.",
+        description="The number of particles to be used in the optimization.",
+    )
+
+    n_iter: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of iterations.",
+        description="The maxmium number of iterations to run the optimization.",
+    )
+
+    cognitive_coeff: Optional[pd.PositiveFloat] = pd.Field(
+        default=1.5,
+        title="Cognitive coefficient.",
+        description="The cognitive parameter decides how attracted the particle is to its previous best position.",
+    )
+
+    social_coeff: Optional[pd.PositiveFloat] = pd.Field(
+        default=1.5,
+        title="Social coefficient.",
+        description="The social parameter decides how attracted the particle is to the global best position found by the swarm.",
+    )
+
+    weight: Optional[pd.PositiveFloat] = pd.Field(
+        default=0.9,
+        title="Weight.",
+        description="The weight or inertia of particles in the optimization.",
+    )
+
+    ftol: Optional[Union[pd.confloat(ge=0, le=1), Literal[-inf]]] = pd.Field(
```

Comment:
same comments about Optional. my understanding is that optional means something can be None. not necessarily that it has a default. Or to put another way, if something has a default, it doesn't need to be Optional, unless None is also an allowed value.

Reply:
Thanks I hadn't realised this. Stripped out all the Optionals now.